getting_started_advanced.md

```diff
@@ -0,0 +1,276 @@
+
+
+### Port Arrays - Definition
+Next, because we will have _n_ LEDs, we will need _n_ IO pins.
+While we can't define a separate top-level IO for each LED, we can define a port array IO that is dynamically sized:  
+```python
+class LedArray(Block):
+  @init_in_parent
+  def __init__(self, count: IntLike) -> None:
+    super().__init__()
+    self.ios = self.Port(Vector(DigitalSink.empty()), [Input])
+    self.gnd = self.Port(Ground.empty(), [Common])
+```
+
+Port arrays are a container port that contain individual ports internally.
+As we saw from Part 1, viewed externally, they have no defined size and ports can be requested from them with optional names.
+However, from internally, we will be able to define their size, as well as get the names of incoming connections.
+
+> Port arrays require the port type to be undefined.
+> Since we have not defined any ports so far, this is only needed for the type and may not have additional data like parameter values.
+
+
+## Implementation
+_In this section, we'll actually implement the circuit generator._
+
+### Generator
+First, we will need a way to get the concrete (`int`) value for the LED count.
+Generators are a way to defer the implementation of the block until its parameter values are ready, then get the concrete Python version of that parameter for use in the HDL.
+**We'll change the base class in our prior code from `Block` to `GeneratorBlock` which provides this functionality**.
+```python
+class LedArray(GeneratorBlock):
+  @init_in_parent
+  def __init__(self, count: IntLike) -> None:
+    super().__init__()
+    self.ios = self.Port(Vector(DigitalSink.empty()), [Input])
+    self.gnd = self.Port(Ground.empty(), [Common])
+    self.generator(self.generate, count)
+    
+  def generate(self, count: int) -> None:
+    ...
+```
+
+> While here we use generators as a way to get a concrete value for circuit generation (the LED count), generators can also be used to do calculations beyond the operations available with the parameters.
+> For example, while we can add two `IntExpr`s (which produces another `IntExpr`), something more complex like square root is not provided.
+> For those cases, use a generator to get the parameter's value, where you have access to the full power of Python.
+
+### Port Arrays - Internal
+So far, the port array is still empty, so we must define its elements.
+With the count available as an int, we can use the `for` loop structure from before:
+```python
+class LedArray(GeneratorBlock):
+  ...
+  def generate(self, count: int) -> None:
+    for i in range(count):
+      self.ios.append_elt(DigitalSink.empty())
+      ...
+```
+
+> Port array's `.append_elt(...)` takes in the same arguments as `self.Port(...)`.
+> 
+> To mark a port array as explicitly having no elements, use `.defined()`.
```

Comment:
maybe there can be an example of how this is used

Reply:
good idea, added